### PR TITLE
Better Forma handling (fix #4291)

### LINKED ISSUE #4291: Kobo Forma support
* KOReader version:
koreader-kobo-arm-linux-gnueabihf-v2015.11-stable
(launch with KFMon-v1.2.6-1-g9ab7d3a)
* Device:
Kobo Forma
(firmware version: N782890001425,4.1.15,4.11.11911,4.1.15,4.1.15,00000000-0000-0000-0000-000000000377)

#### Issue
In new released Kobo Forma, when launching Koreader, it crashed right away, reporting "unrecognized Kobo model frost".

Complete detailed log:
[crash.log](https://github.com/koreader/koreader/files/2523197/crash.log)

#### Steps to reproduce
After installed, try to launch Koread, it will crash with above crash log.

##### `crash.log` (if applicable)
./luajit: frontend/device/kobo/device.lua:177: unrecognized Kobo model frost
stack traceback:
	[C]: in function 'error'
	frontend/device/kobo/device.lua:177: in main chunk
	[C]: in function 'probeDevice'
	frontend/device.lua:37: in main chunk
	[C]: in function 'require'
	frontend/ui/uimanager.lua:1: in main chunk
	[C]: in function 'require'
	./reader.lua:92: in main chunk
	[C]: at 0x0000bea5

### PR DESCRIPTION
**Pull request by NiLuJe**
* Enforce Portrait on startup, to avoid getting touch coordinates wrong if we were started from Nickel in inverted Portrait.
* Fix button mapping, and made it behave properly in landscape. I may have massively ruined non-Portrait mappings for everyone else, but, hey, this made sense, and it works here ;p.
* Fix NaturalLight handling, thanks to @cairnsh & @pazos in #4291 ;)

### CODE REVIEW

**Comment by NiLuJe** [comment]
The oly midly annoying thing left is the FrontLight widget: it expects a scale of 0 to 100 for the warmth, while this device only actually handles 0 to 10, so this leaves a lot of "dead" steps.

UI work is really not my thing, so I'm leaving that as an exercise to the reader for now.
(The fact that I personally barely ever move from a single warmth step doesn't help).

**Comment by NiLuJe** [comment]
Waiting on [BASE#780](https://github.com/koreader/koreader-base/pull/780) ;).

**Comment by Frenzie** [comment]
Completely off topic, but would you mind taking a quick peek on [Gitter](https://gitter.im/koreader/koreader?at=5c1d7f6bdb5b5c6883e9c2b3)? I don't know the Kindle details. ;-)

**Comment by NiLuJe** [comment]
Random remark about the button remapping on rotation:

I just checked on old Kindles with buttons (K3 & K4), and on landscape, they do *NOT* remap the page turn buttons, which is stupid.

The current status of the code seems to come from the Oasis days, and those *may* do funky stuff with button keycodes depending on orientation, and/or having the buttons bound in the reverse to begin with.

So, yeah, this makes sense on the Forma & on non-touch Kindles, but it may feel weird on the Oasis 1 or 2, or both.

**Comment by NiLuJe** [comment]
I'm also hoping the Nickel setting to invert the buttons is entirely handled on Nickel's side and doesn't actually affect us on the Forma ;).

**Comment by NiLuJe** [comment]
Speaking of UI stuff, being able to switch to inverted rotations would be NTH (especially inverted portrait here), but the ScreenMode stuff seems fairly gnarly, so, err, pass ;p.

(Besides adding more buttons to the bottom menu, I was thinking a tap on the orientation icon would swap the icon to a mirrored version, and the button choices to their Inverted variant, with an immediate toggle to the Inverted variant of the current orientation).

**Comment by NiLuJe** [comment]
So, on my end, that's it for now, ready to go :).

**Comment by Frenzie** [comment]
If you tap the bottom orientation thingy twice then you actually get inverted, don't you?

**Comment by NiLuJe** [comment]
@Frenzie : Doesn't appear to?

**Comment by Frenzie** [comment]
Might be that it was never implemented for portrait, only for landscape.

**Comment by NiLuJe** [comment]
Ah! Indeed, that works for Landscape!

I'll see what I can do...

File: frontend/device/sysfs_light.lua [hunk 1]
```diff
@@ -56,29 +57,38 @@ function SysfsLight:setNaturalBrightness(brightness, warmth)
...
-    local white = 0
-    if brightness > 0 then
-        -- On Nickel, the values for white/red/green are roughly linearly dependent
-        -- on the 4th root of brightness and warmth.
-        white = math.min(self.white_gain * math.pow(brightness, self.exponent) *
+    -- Newer devices use a mixer instead of writting values per color.
+    if self.frontlight_mixer then
+        -- And the warmth is now on a scale of 0 to 10 ;).
+        warmth = math.floor(warmth/10)
+        self:_write_value(self.frontlight_white, brightness)
```
Comment (pazos) [inline frontend/device/sysfs_light.lua]: can we use `self:_set_light_value(self.frontlight_white, brightness)` and leave frontlight_white as is?

File: frontend/device/kobo/device.lua [hunk 2]
```diff
@@ -199,23 +199,22 @@ local KoboNova = Kobo:new{
...
     hasFrontlight = yes,
     hasKeys = yes,
-    touch_probe_ev_epoch_time = true,
     touch_snow_protocol = true,
     display_dpi = 300,
     hasNaturalLight = yes,
     frontlight_settings = {
-        frontlight_white = "/sys/class/backlight/mxc_msp430.0",
-        frontlight_red = "/sys/class/backlight/tlc5947_bl",
+        frontlight_white = "/sys/class/backlight/mxc_msp430.0/brightness",
```
Comment (pazos) [inline frontend/device/kobo/device.lua]: see the other comment :+1:

File: frontend/device/sysfs_light.lua (hunk 1 above)
Comment (pazos) [reply frontend/device/sysfs_light.lua]: I mean, writting bl_power is not neccesary but shouldn't hurt.

Comment (NiLuJe) [reply frontend/device/sysfs_light.lua]: `-1676681792`, to be exact ;p.

```
┌─(ROOT@(none):pts/0)───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(/)─┐
└─(0.22:58%:22:28:96%:#)── cat /sys/class/backlight/mxc_msp430.0/bl_power                                                                                                                                                                                                                                                                               ──(Sat, Dec 22)─┘
-1676681792
┌─(ROOT@(none):pts/0)───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(/)─┐
└─(0.20:58%:22:28:96%:#)── cat /sys/class/backlight/tlc5947_bl/bl_power                                                                                                                                                                                                                                                                                 ──(Sat, Dec 22)─┘
0
```

Comment (NiLuJe) [reply frontend/device/sysfs_light.lua]: Given that, I refrained from trying to write to it ;p.

**Comment by NiLuJe** [comment]
Okay, I *think* I got the rotation logic right, and fixed an old bug in the process (maybe?).

This depends on [BASE#781](https://github.com/koreader/koreader-base/pull/781).

It's a bit weird, I couldn't get ReaderView:onSetScreenMode to take three args (ideally, `new_mode, rotation, interactive`), since I'd to set rotation to nil in ReaderRolling:onChangeScreenMode: doing `self.ui:handleEvent(Event:new("SetScreenMode", mode, nil, true))` in there appeared to drop that true?

----

In the same vein, this probably doesn't handle PDF, because koptoptions uses the SetScreenMode event directly, and I couldn't figure out how to pass it an extra argument that isn't part of the options...

**Comment by NiLuJe** [comment]
@pazos: That's an interesting comment about the gyro.... Not familiar with this, what should I check?

**Comment by pazos** [comment]
If the gyro is handled by udev rules I guess that you can find a string in /tmp/nickel-hardware-status after rotation (and the rule somewhere in /usr/local/Kobo).

If there isn't a specific rule for the gyro then I guess they're polling some device under `/sys/devices/platform`. or reading the value from /dev/ntx_io.

In any case the chip would be something like [MMA8491Q](https://www.nxp.com/docs/en/data-sheet/MMA8491Q.pdf) and accesible via I2C interface.

**Comment by pazos** [comment]
Yeah, Forma kernel has `CONFIG_SENSORS_MMA8X5X=y` so yep.

**Comment by NiLuJe** [comment]
Oh, `/dev/input/event2` is *definitely* the gyro, it's verbose as hell, but I have no idea how to parse that :D.

**Comment by NiLuJe** [comment]
`mma8652`, FWIW.

**Comment by pazos** [comment]
@NiLuJe evtest is your friend

**Comment by NiLuJe** [comment]
Noooo, not input events xD.

So,
ABS_X/ABS_Y/ABS_Z (with Z being useless), and x/y following some sort of logic depending on the rotation.

UR: X ~= -16000 & Y ~= 0
CW: X ~= 0 & Y ~= -16000
UD: X ~= -16000 & Y ~= 0
CCW: X ~= 0 & Y ~= -16000

So, you need to keep track of this thing (because it's an accelerometer, not a true gyro?).

**Comment by pazos** [comment]
From [stackexchange](https://unix.stackexchange.com/questions/335260/accelerometer-screen-rotation-on-non-touchscreen-laptop/335516) , other chip but same interface

**/proc/bus/input/devices**

```
    I: Bus=0019 Vendor=0000 Product=0000 Version=0000
    N: Name="ST LIS3LV02DL Accelerometer"
    P: Phys=lis3lv02d/input0
    S: Sysfs=/devices/platform/lis3lv02d/input/input7
    U: Uniq=
    H: Handlers=event6 js0
    B: PROP=0
    B: EV=9
    B: ABS=7
```
**evtest /dev/input/event6**
```
    Input device ID: bus 0x19 vendor 0x0 product 0x0 version 0x0
    Input device name: "ST LIS3LV02DL Accelerometer"
    Supported events:
      Event type 0 (EV_SYN)
      Event type 3 (EV_ABS)
        Event code 0 (ABS_X)
          Value     20
          Min    -2304
          Max     2304
          Fuzz      18
          Flat      18
        Event code 1 (ABS_Y)
          Value    -38
          Min    -2304
          Max     2304
          Fuzz      18
          Flat      18
        Event code 2 (ABS_Z)
          Value   1105
          Min    -2304
          Max     2304
          Fuzz      18
          Flat      18
    Properties:
    Testing ... (interrupt to exit)
```

**Comment by NiLuJe** [comment]
Okay, yeah, [Z might not be useless](https://github.com/koreader/koreader/pull/4414#issuecomment-449605323), that was just Kobo's evtest build being contrary.

**Comment by pazos** [comment]
@NiLuJe: I think we can repurpose most orientation code on https://github.com/koreader/koreader/blob/master/frontend/device/input.lua.

On oasis there is ABS_X, ABS_Y and ABS_PRESSURE. Here we have  ABS_X, ABS_Y and ABS_Z.

**Comment by NiLuJe** [comment]
Fun fact: nothing appears to be holding an fd to event2 while nickel is running, so there may be another (hopefully easier to parse) way to get the orientation info? `ntx_io` maybe?

**Comment by Frenzie** [comment]
>  but doing so, self.ui:handleEvent(Event:new("SetScreenMode", mode, nil, true)) appeared to drop that true once it got back to ReaderView:onSetScreenMode (but rotation being nil indeed made it through

That should work… Stupid question, but did you grep the whole source for SetScreenMode?

**Comment by NiLuJe** [comment]
Yep, those are the only instance of `onChangeScreenMode` and `onSetScreenMode`, and the only instance of `setScreenMode` ;).

**Comment by NiLuJe** [comment]
Might be a bit of Lua + vararg trickery...

https://stackoverflow.com/questions/7183998/in-lua-what-is-the-right-way-to-handle-varargs-which-contains-nil

**Comment by NiLuJe** [comment]
Meaning `args = { n = select("#", ...), ... }` instead of `args = {...}` in `Event:new`'s body would be enough to do the trick?

**Comment by NiLuJe** [comment]
Wait, no, that made no sense xD.

I'm open to suggestions, because brainhurty :D.

**Comment by Frenzie** [comment]
Oh, I didn't read that closely enough, my bad. Yeah, never mind for now, lol. It's clear enough with the comments in base. :-)

**Comment by pazos** [comment]
> I dunno, to me that reads like the Oasis accelerometer was spouting ready-to-consume codes like "I"m in this rotation" (from ABS_PRESSURE).

> Here, I'm getting the raw X/Y/Z values, updated every 50ms!

You're right! We can generate orientation codes from raw values anyways, but before ...

> Fun fact: nothing appears to be holding an fd to event2 while nickel is running, so there may be another (hopefully easier to parse) way to get the orientation info?

CM_ROTARY_STATUS has been in the firmware for ages, without usage. On forma kernel `under arch/arm/mach-imx/imx_ntx_io.c`

```
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/ioctl.h>
#include <sys/stat.h>

#define CM_ROTARY_STATUS 200

int main(int argc,char **argv)
{
 int ntx_io;
 unsigned long ret,pointer,value;
 ntx_io=open("/dev/ntx_io",O_RDWR);
 retval=ioctl(ntx_io, CM_ROTARY_STATUS, &pointer);
 printf("ioctl(ntx_io,CM_ROTARY_STATUS) = %d; returned %ld\n",ret,pointer);
 close(ntx_io);
}
```

**Comment by NiLuJe** [comment]
@pazos: No such luck :/

```
ioctl(ntx_io, CM_ROTARY_STATUS) = 0; returned 0
```

Both CM_ROTARY_* commands seem massively commented out/empty in the Forma kernel, and I can't sniff nickel doing anything with ntx_io related to rotation (it does use it to set the FL, while using sysfs to set the color mixer...).

File: frontend/device/kobo/powerd.lua [hunk 3]
```diff
@@ -112,8 +112,10 @@ function KoboPowerD:init()
     self.autowarmth_job_running = false
 
     if self.device.hasFrontlight() then
-        -- If this device has natural light (currently only KA1)
-        -- use the SysFS interface, and ioctl otherwise.
+        -- If this device has natural light (currently only KA1 & Forma)
+        -- Use the SysFS interface, and ioctl otherwise.
+        -- NOTE: On the Forma, nickel still appear to prefer using ntx_io to handle the FL,
```
Comment (Frenzie) [inline frontend/device/kobo/powerd.lua]: :rofl:

Comment (NiLuJe) [reply frontend/device/kobo/powerd.lua]: Yeah, I'm starting to believe the blind monkeys may have escaped from the QA department, and managed to infiltrate the dev team, too :D.

Comment (NiLuJe) [reply frontend/device/kobo/powerd.lua]: :see_no_evil: :hear_no_evil: :speak_no_evil:

Comment (NiLuJe) [reply frontend/device/kobo/powerd.lua]: Snark aside, it's impressive how sharp the contrast is between Kobo & lab126 via the glimpse into the development process we get by tinkering with this stuff...

Granted, lab126 can probably afford to throw walls of money at the problem ;).

Comment (Frenzie) [reply frontend/device/kobo/powerd.lua]: Although at least some of the lack of… finesse is for the better. ;-)

Comment (NiLuJe) [reply frontend/device/kobo/powerd.lua]: Indeed, and let's not forget the generally not being evil part, too ;).

Comment (NiLuJe) [reply frontend/device/kobo/powerd.lua]: Basically, that's just my yearly realization that Amazon still has Amazon problems, and Kobo still has Kobo problems, what was true 8 years ago is mostly still true today ;).

**Comment by NiLuJe** [comment]
Mmkay, after a look at `drivers/hwmon/mma8x5x.c`, the orientation changed events *should* end up getting thrown to event0...

**Comment by NiLuJe** [comment]
`include/uapi/linux/input.h `
```C
#define EV_MSC                  0x04

#define MSC_RAW_GSENSOR_PORTRAIT_DOWN   0x17
#define MSC_RAW_GSENSOR_PORTRAIT_UP             0x18
#define MSC_RAW_GSENSOR_LANDSCAPE_RIGHT 0x19
#define MSC_RAW_GSENSOR_LANDSCAPE_LEFT  0x1a
#define MSC_RAW_GSENSOR_BACK                    0x1b
#define MSC_RAW_GSENSOR_FRONT                   0x1c
```

```
12/23/18-21:04:17 DEBUG input event => type: 4, code: 3(nil), value: 24, time: 1545595457.13777
12/23/18-21:04:17 DEBUG input event => type: 0, code: 0(nil), value: 0, time: 1545595457.13777
12/23/18-21:04:17 DEBUG input event => type: 4, code: 3(nil), value: 28, time: 1545595457.15236
12/23/18-21:04:17 DEBUG input event => type: 0, code: 0(nil), value: 0, time: 1545595457.15236
```

```
12/23/18-21:04:35 DEBUG input event => type: 4, code: 3(nil), value: 25, time: 1545595475.106305
12/23/18-21:04:35 DEBUG input event => type: 0, code: 0(nil), value: 0, time: 1545595475.106305
12/23/18-21:04:35 DEBUG input event => type: 4, code: 3(nil), value: 28, time: 1545595475.106343
12/23/18-21:04:35 DEBUG input event => type: 0, code: 0(nil), value: 0, time: 1545595475.106343
```

```
12/23/18-21:04:49 DEBUG input event => type: 4, code: 3(nil), value: 23, time: 1545595489.533658
12/23/18-21:04:49 DEBUG input event => type: 0, code: 0(nil), value: 0, time: 1545595489.533658
12/23/18-21:04:49 DEBUG input event => type: 4, code: 3(nil), value: 28, time: 1545595489.535211
12/23/18-21:04:49 DEBUG input event => type: 0, code: 0(nil), value: 0, time: 1545595489.535211
```

```
12/23/18-21:05:18 DEBUG input event => type: 4, code: 3(nil), value: 26, time: 1545595518.106326
12/23/18-21:05:18 DEBUG input event => type: 0, code: 0(nil), value: 0, time: 1545595518.106326
12/23/18-21:05:18 DEBUG input event => type: 4, code: 3(nil), value: 28, time: 1545595518.106363
12/23/18-21:05:18 DEBUG input event => type: 0, code: 0(nil), value: 0, time: 1545595518.106363
```

**Comment by NiLuJe** [comment]
As for BACK & FRONT, that indeed matches whether the device is face-up or face-down ;).

**Comment by NiLuJe** [comment]
:pizza: break, I'll see if I can cobble something up later ;).

**Comment by NiLuJe** [comment]
\o/

**Comment by NiLuJe** [comment]
Fun fact: this one appears to go to deep sleep even while plugged to a computer/in USBNet mode!

My connection dropped while I was out after KOReader's auto-sleep, something which *never* happened on the H2O!

```
[Dec23 20:57] usb 1-10: new high-speed USB device number 19 using xhci_hcd
[  +0.130189] cdc_ether 1-10:1.0 usb0: register 'cdc_ether' at usb-0000:00:14.0-10, CDC Ethernet Device, ee:49:00:00:00:00
[  +0.011951] cdc_ether 1-10:1.0 enp0s20f0u10: renamed from usb0
[Dec23 22:00] usb 1-10: USB disconnect, device number 19
[  +0.000048] cdc_ether 1-10:1.0 enp0s20f0u10: unregister 'cdc_ether' usb-0000:00:14.0-10, CDC Ethernet Device
[Dec23 23:15] usb 1-10: new high-speed USB device number 20 using xhci_hcd
[  +0.134536] cdc_ether 1-10:1.0 usb0: register 'cdc_ether' at usb-0000:00:14.0-10, CDC Ethernet Device, ee:49:00:00:00:00
[  +0.003399] cdc_ether 1-10:1.0 enp0s20f0u10: renamed from usb0
```

File: frontend/device/kobo/powerd.lua (hunk 3 above)
Comment (pazos) [reply frontend/device/kobo/powerd.lua]: I would love to see nickel hardware abstraction layer. 

lab126 is other league. They have the ['guy' who ran linux on a 8-bit micro](http://dmitry.gr/?r=05.Projects&proj=07.%20Linux%20on%208bit)

**Comment by NiLuJe** [comment]
Oohkay, I'm done! :).

**Comment by NiLuJe** [comment]
If a ColorSetting that falls *between* the 11 actual breakpoints is stored by KOReader, Nickel will round it to the nearest step, so, not much harm done, besides a potential 1-step diff that can/will easily be corrected.

**Comment by pazos** [comment]
@NiLuJe: I guess that other devices without red leds will be ok using the new mixer interface (Since the current behaviour leds to slightly modified white brightness when applying a new color). I'm talking about Kobo Clara HD and Aura H2O2, Rev2. In both cases the frontlight.mixer is /sys/class/backlight/lm3630a_led/color

**Comment by NiLuJe** [comment]
Possibly, but I don't have any of those to check if the scale is as finicky as this one, and apparently it works as it currently is, so, eh ;).

**Comment by NiLuJe** [comment]
In any case, that reminded me of something I'd forgotten to jot down on my TODO-list, so, at least if someone wants to try, they easily can since the scale is no longer hard-coded ;).

**Comment by NiLuJe** [comment]
Final request for comment, @Frenzie / @pazos , before I merge this?

**Comment by faassen** [comment]
I'm a random guy who may not know what he's doing with a new kobo forma: I built this branch locally (with koxtoolchain; at least I hope -- it's installed and presumably used when I use './kodev release kobo'). 

Then I used kfmon to start it. Unfortunately when I try to start koreader it crashes immediately (some message about it crashing flashes by quickly) and then reboots the kobo forma. I tried using the proper pre-built release but that was even worse: the kobo forma crashed and presumably had a broken database as it didn't pick up any books anymore, but that might have just been a case of extra bad luck.

Is there any way I can help in debugging this?

**Comment by Frenzie** [comment]
@NiLuJe 
Nothing jumps out at a quick skim.

@faassen 
> I built this branch locally (with koxtoolchain; at least I hope -- it's installed and presumably used when I use './kodev release kobo').

It would be if it's in PATH, otherwise likely not. It says which compiler it's using. But note that this PR only changes some Lua code, so besides making sure you also update the relevant file in base no compilation is necessary.

> I tried using the proper pre-built release but that was even worse: the kobo forma crashed and presumably had a broken database as it didn't pick up any books anymore, but that might have just been a case of extra bad luck.

How did you install and run it?

**Comment by faassen** [comment]
I don't want to hold up this PR and this was definitely in the wrong place to comment, my apologies. It's just I went out of my way to try to build this branch specifically. I'll move over to an issue.

**Comment by NiLuJe** [comment]
As @Frenzie said, there's no compiled code changes, so, no need to build anything ;).

The way I personally do it is start from the latest nightly, then a simple

```
for file in $(git diff --name-only upstream/master) ; do ; [[ -f "${file}" ]] && scp "${file}" "root@kindle:/mnt/onboard/.adds/koreader/${file}" ; done
```

In both my koreader & koreader-base branch.

(That works for mostly everything, except the stuff in platform/, so in this specific case, I scp'ed that manually).

(And, yes, *all* my devices are aliased to "kindle" over USBNet, ^^).

**Comment by faassen** [comment]
Thanks for the tip! I will try that in the future.

Here's the issue. https://github.com/koreader/koreader/issues/4425

I am looking forward to trying out your work on my Forma!

**Pull request merged**